### PR TITLE
Allows plain body to be set, rather than just JSON or Form variables

### PR DESCRIPTION
**Pull request by andyjeffries**
Hi Dalton,

I wanted to use https://github.com/Jeffail/gabs with Sling, but found that you don't allow setting the body manually, only from an existing JSON-marshallable structure or form data.  Here's a PR that allows you to manually set the body.

I'm pretty new to Go, so if I've done anything daft feel free to let me know and I'll amend it.

Cheers,

Andy

### CODE REVIEW

**Comment by dghubble** [comment]
Hey thanks for the PR. 

Sling separates building a request (with various boilerplate setters, culminating in a call to `Request() (*http.Request, error)`) and executing the request (with `Do(req *http.Request, success, failure interface{})`) is to allow developers to make further modifications to the Request before executing.

You can already set a plain body on the http.Request directly.

    req, err := sling.New().Base("http://example.com/").Path("foo").Request()
    req.Body = ioutil.NopCloser(strings.NewReader(expectedBody))

You're right though, there is a reasonable case for a `Body` setter to compliment `BodyForm` and `BodyJSON` so the `Receive` and `ReceiveSuccess` convenience methods can be used.

However, its signature should be `func (*Sling) Body(body io.Reader) *Sling` because a string body is just one particular possibility. `http.NewRequest` is a useful [example](http://golang.org/src/net/http/request.go?s=17718:17790#L536)

**Comment by andyjeffries** [comment]
Thanks for the helpful reply/comment.

I've changed the commit and pushed the new version with the correct signature.  Is that what you require?  Have I done it the right way?

**Comment by andyjeffries** [comment]
Actually the docs say that `strings.NewReader` is more efficient than `bytes.NewBufferString`, but for the test I guess this doesn't matter.  I got `bytes.NewBufferString` from a mailing list post as the recommend way, but if you want me to change it, let me know.

File: sling.go [hunk 1]
```diff
@@ -33,6 +33,8 @@ type Sling struct {
 	bodyJSON interface{}
 	// url tagged body struct (form)
 	bodyForm interface{}
+	// simply assigned body
+	body string
```
Comment (dghubble) [inline sling.go]: // plain Body
body io.ReadCloser

File: sling.go [hunk 2]
```diff
@@ -200,6 +203,18 @@ func (s *Sling) BodyForm(bodyForm interface{}) *Sling {
 	return s
 }
 
+// Body sets the Sling's body. The value pointed to by the
+// body will be set raw/as-is as the Body on new requests.
+// The body argument should be a string
+func (s *Sling) Body(r io.Reader) *Sling {
+	body := new(bytes.Buffer)
```
Comment (dghubble) [inline sling.go]: Probably something more like:

    func (s *Sling) Body(body io.Reader) *Sling {
        rc, ok := body.(io.ReadCloser)
        if !ok && body != nil {
            rc = ioutil.NopCloser(body)
        }
        s.body = rc
        return s
    }

File: sling.go [hunk 3]
```diff
@@ -264,6 +279,8 @@ func (s *Sling) getRequestBody() (body io.Reader, err error) {
 		if err != nil {
 			return nil, err
 		}
+	} else if s.body != "" {
```
Comment (dghubble) [inline sling.go]: compare against nil and set body to be s.body

File: sling_test.go [hunk 4]
```diff
@@ -403,6 +435,8 @@ func TestRequest_body(t *testing.T) {
 		{New().BodyJSON(modelA).New().BodyForm(paramsB), "count=25&kind_name=recent", formContentType},
 		{New().BodyForm(paramsB).New().BodyJSON(nil), "count=25&kind_name=recent", formContentType},
 		{New().BodyJSON(modelA).New().BodyForm(nil), "{\"text\":\"note\",\"favorite_count\":12}\n", jsonContentType},
+		// Body
+		{New().Body(bytes.NewBufferString("this-is-a-test")), "this-is-a-test", ""},
```
Comment (dghubble) [inline sling_test.go]: yes, just use strings.NewReader in your test

**Comment by andyjeffries** [comment]
Thanks for the Dalton, I appreciate you taking the time to help me make this better.  Let's see if this version is OK?  I think `TestBodySetter` could be better (I'm not a fan of the empty string switching it with nil), but I couldn't figure out a better way.

**Comment by dghubble** [comment]
Ah, I see the problem. If we make a separate 

    var testInput = ioutil.NopCloser(strings.NewReader("test"))

 and use `io.Reader` or `io.ReadCloser` directly, the empty string switching can be avoided.

I tweaked that little bit and merged as 3b289d05ceacdc5e4ae2fbdcde7ea4d5d67ccfdf Thanks! Hope you're enjoying Go!

**Pull request closed without merging**